src/server/worker/datum/iterator_v2.go

```diff
@@ -0,0 +1,386 @@
+//	if input.Commit == "" {
+//		// this can happen if a pipeline with multiple inputs has been triggered
+//		// before all commits have inputs
+//		return &gitIteratorV2{}, nil
+//	}
+//	fi, err := pachClient.InspectFileV2(input.Name, input.Commit, "/commit.json")
+//	if err != nil {
+//		return nil, err
+//	}
+//	return &gitIteratorV2{
+//		inputs: []*common.Input{
+//			&common.Input{
+//				FileInfo: fi,
+//				Name:     input.Name,
+//				Branch:   input.Branch,
+//				GitURL:   input.URL,
+//			},
+//		},
+//	}, nil
+//}
+//
+//func (gi *gitIteratorV2) Iterate(cb func([]*common.Input) error) error {
+//	if len(gi.inputs) == 0 {
+//		return nil
+//	}
+//	return cb(gi.inputs)
+//}
+
+func newCronIteratorV2(pachClient *client.APIClient, input *pps.CronInput) IteratorV2 {
+	return newPFSIteratorV2(pachClient, &pps.PFSInput{
+		Name:   input.Name,
+		Repo:   input.Repo,
+		Branch: "master",
+		Commit: input.Commit,
+		Glob:   "/*",
+	})
+}
+
+// Hasher is the standard interface for a datum hasher.
+type Hasher interface {
+	// Hash computes the datum hash based on the inputs.
+	Hash([]*common.Input) string
+}
+
+type jobIterator struct {
+	iterator IteratorV2
+	jobID    string
+	hasher   Hasher
+}
+
+// NewJobIterator creates a new job iterator.
+func NewJobIterator(iterator IteratorV2, jobID string, hasher Hasher) IteratorV2 {
+	return &jobIterator{
+		iterator: iterator,
+		jobID:    jobID,
+		hasher:   hasher,
+	}
+}
+
+func (ji *jobIterator) Iterate(cb func(*Meta) error) error {
+	return ji.iterator.Iterate(func(meta *Meta) error {
+		meta.JobID = ji.jobID
+		meta.Hash = ji.hasher.Hash(meta.Inputs)
+		return cb(meta)
+	})
+}
+
+type fileSetIterator struct {
+	pachClient   *client.APIClient
+	repo, commit string
+}
+
+// NewFileSetIterator creates a new fileset iterator.
+func NewFileSetIterator(pachClient *client.APIClient, repo, commit string) IteratorV2 {
+	return &fileSetIterator{
+		pachClient: pachClient,
+		repo:       repo,
+		commit:     commit,
+	}
+}
+
+func (fsi *fileSetIterator) Iterate(cb func(*Meta) error) error {
+	r, err := fsi.pachClient.GetTarV2(fsi.repo, fsi.commit, path.Join("/", MetaPrefix, "*", MetaFileName))
+	if err != nil {
+		return err
+	}
+	tr := tar.NewReader(r)
+	for {
+		_, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+		meta := &Meta{}
+		if err := jsonpb.Unmarshal(tr, meta); err != nil {
+			return err
+		}
+		if err := cb(meta); err != nil {
+			return err
+		}
+	}
+}
+
+// Merge merges multiple datum iterators (key is datum ID).
+func Merge(dits []IteratorV2, cb func([]*Meta) error) error {
+	var ss []stream.Stream
+	for _, dit := range dits {
+		ss = append(ss, newDatumStream(dit, len(ss)))
+	}
+	pq := stream.NewPriorityQueue(ss)
+	return pq.Iterate(func(ss []stream.Stream, _ ...string) error {
+		var metas []*Meta
+		for _, s := range ss {
+			metas = append(metas, s.(*datumStream).meta)
```

Comment:
Yeah, it is because of the priority queue interface which has sort of shape shifted over time. I will probably circle back to this at some point to tighten up the streaming priority queue interface, but it does not seem like a high priority right now.